pkg/storage/multi/saver.go

```diff
@@ -0,0 +1,50 @@
+package multi
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+	"github.com/gomods/athens/pkg/storage"
+)
+
+// Save implements the (github.com/gomods/athens/pkg/storage).Saver interface.
+func (s *Storage) Save(ctx context.Context, module, version string, mod []byte, zip io.Reader, info []byte) error {
+	const op errors.Op = "mutli.Delete"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+
+	content, err := ioutil.ReadAll(zip)
```

Comment:
yes, this is the bottleneck here, i need to think this through, but i was out of ideas.
was thinking about combining teereader and multiwriter which would then write to readers passed down the stack, but this would not work as well

Reply:
thought about this today for few hours. I came up with this
https://medium.com/@MichalPristas/forwarding-reader-to-multiple-consumers-a1574a9db4d6
but i dont like it 😄

Reply:
I didn't try this out but maybe something like this? For simplicity let's say we have two storages:
```
...
pr1, pw1 := io.Pipe()
pr2, pw2 := io.Pipe()

var results = make(chan error, 2)
defer close(results)

go func() { 
  results <- storage1.Save(c, module, version, mod, pr1, info)
}
go func() { 
  results <- storage2.Save(c, module, version, mod, pr2, info)
}

go func() {
    defer pw1.Close()
    defer pw2.Close()

    mw := io.MultiWriter(pw1, pw2)
    io.Copy(mw, zip) // copy from zip into multiwriter- this should fan out into pr1 and pr2
}()
for i := 0; i < 2; i++ {
	r := <-results
	if r != nil {
		errs = append(errs, r)
	}
}
```
Let me know what you think.

Reply:
I will try the copy out. should work just like with tee writer but with less code. I like that. still will be as fast as the slowest reader though.

Reply:
I checked the implementation of io bits, seems that Copy uses buffer internally of size 32*1024 bytes.
Using tee + pipe is byte by byte so lower memory footprint. 32k is not a huge amount of memory and i think it would do ok.